Confirm WRITE and CREATE failures for automated tests

xBis7 opened the pull request.

xBis7 commented on trino-hms-hdfs-ranger/big-data-c3-tests/lib.sh:
```diff
@@ -268,9 +268,11 @@ runTrino() {
-    docker exec -it -u "$user" "$TRINO_HOSTNAME" trino --execute="$trino_cmd" 2>&1 | tee "$TRINO_TMP_OUTPUT_FILE"
+    # Use '--output-format ALIGNED' to make sure that there is always output and it has the correct format.
+    # This option, aligns the output columns in tabs.
+    docker exec -it -u "$user" "$TRINO_HOSTNAME" trino --output-format ALIGNED --execute="$trino_cmd" 2>&1 | tee "$TRINO_TMP_OUTPUT_FILE"
```
`--output-format ALIGNED` This addition is because in a query that doesn't return a result and the output is `(0 rows)`, it doesn't print anything.

If we run the command directly in trino-shell then it's ok, but there is no output when using `docker exec`. This fixes it but I don't have a way of confirming whether this works in c3 as well.

xBis7 commented:
I'm looking into better error handling in case of failure. This is some example output where the test failed.

```
=> Testing that 'createTable' has failed as expected.

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/07/18 11:05:47 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://spark-master:4040
Spark context available as 'sc' (master = spark://spark-master:7077, app id = app-20240718110547-0002).
Spark session available as 'spark'.


Running command:
spark.sql("select * from gross_test.test2").show

Expecting it to fail.
--------------------------


Hive Session ID = 90916fe8-3e04-4d47-957e-a6ad979a492e
+---+----+                                                                      
| id|name|
+---+----+
|  1|John|
|  2|Jane|
|  3| Bob|
+---+----+



--------------------------
Test finished without issues while it was expected to fail.
--------------------------



=> There was an error. 'createTable' was expected to have failed but it was successful.
```

I'll run all the tests again and then push the changes.

xBis7 commented:
All tests passed. I've updated the PR.

GeorgeJahad commented on trino-hms-hdfs-ranger/big-data-c3-tests/lib.sh:
```diff
@@ -293,6 +295,131 @@ runTrino() {
 
+# -- CHECK CREATE/WRITE FAILURES --
+verifyCreateWriteFailure() {
+  # Run all checks as user1 for now.
```
Is this because user1 has more permissions? If so, then please add that to the comment.

xBis7 replied:
Yes, that's correct. I'll update the comment.

GeorgeJahad commented on trino-hms-hdfs-ranger/big-data-c3-tests/lib.sh:
```diff
@@ -293,6 +295,131 @@ runTrino() {
+# -- CHECK CREATE/WRITE FAILURES --
+verifyCreateWriteFailure() {
+  # Run all checks as user1 for now.
+  component=$1
+  operation=$2
```
Please list the possible values here.  We haven't been doing a very good job of documenting parameters.  Let's start doing a better job of it.

GeorgeJahad commented on trino-hms-hdfs-ranger/big-data-c3-tests/lib.sh:
```diff
@@ -293,6 +295,131 @@ runTrino() {
+  operation=$2
+  db_name=$3
+  table_name=$4
+  insert_id=$5 # All tables have an id field.
```
I know what you mean here, but to me it is clearer to say:
"All rows have an id value passed in during the insert."

also if this is an optional parameter please make that clear.

xBis7 replied:
I've updated the comment.

GeorgeJahad commented on trino-hms-hdfs-ranger/big-data-c3-tests/lib.sh:
```diff
@@ -293,6 +295,131 @@ runTrino() {
+      command="show tables in $TRINO_HIVE_SCHEMA.$db_name"
+      expectedOutput="Schema '$db_name' does not exist"
+
+      runTrino "$TRINO_USER1" "$command" "shouldFail" "$expectedOutput" "user"
```
USER1 shouldn't have the final "user" parameter

xBis7 replied:
That `"user"` parameter has something to do with c3. I'm not sure what is used for but because @mladjan-gadzic added it in the other calls, I added it here as well.

xBis7 replied:
I cleaned it up.

GeorgeJahad commented on trino-hms-hdfs-ranger/big-data-c3-tests/lib.sh:
```diff
@@ -293,6 +295,131 @@ runTrino() {
+      command="show tables in $TRINO_HIVE_SCHEMA.$db_name"
+      expectedOutput="$table_name"
+
+      runTrino "$TRINO_USER1" "$command" "shouldPass" "$expectedOutput" "user"
```
same as above

GeorgeJahad commented:
why are renames not checked?

spark tests 10 and 11 and trino tests 9 and 11 all have alters.

GeorgeJahad commented on trino-hms-hdfs-ranger/big-data-c3-tests/trino-spark-tests/trino/test_9.sh:
```diff
@@ -45,7 +45,10 @@ expectedMsg="Permission denied: user [$TRINO_USER1] does not have [SELECT] privi
+runTrino "$TRINO_USER1" "$command" "shouldFail" "$expectedMsg"
+
+# We can't check if the create failed here because there is no READ access for any user.
+# verifyCreateWriteFailure "trino" "createDb" "gross_test"
```
> # We can't check if the create failed here because there is no READ access for any user.


Can we just change permissions temporarily?

xBis7 replied:
@GeorgeJahad It turns out that although the schema creation should have failed and it gave us an exception, the create worked. `verifyCreateWriteFailure` showed that the schema was actually created.

```
=> Testing that 'createDb' has failed as expected.


==========================
Running trino command: 'show tables in hive.gross_test' for user 'trino' 
--------------------------

 Table 
-------
(0 rows)

--------------------------
The command failed. It didn't have the expected output.
Expected output: 'Schema 'gross_test' does not exist'
Actual output: ' Table 
-------
'0 rows)
==========================
```

I'll look more into this.

xBis7 replied:
I just realized that `test_7.sh` creates the schema. Here we are expecting to get a SELECT error while trying to create a schema that already exists without having enough permissions to read and figure out that it's there.

IMO, we should add a comment here saying that and skip the check.

xBis7 replied:
I added a comment explaining that the db already exists from `test_7.sh`.

xBis7 commented:
> why are renames not checked?
> 
> spark tests 10 and 11 and trino tests 9 and 11 all have alters.

@GeorgeJahad I've extended the code to check for rename failures. I've run all the tests and there are no issues.

xBis7 commented:
@GeorgeJahad I've updated the PR. I'm not sure about the `"user"` parameter in `runTrino`.

GeorgeJahad reviewed (approved):
lgtm

xBis7 commented:
* rebased the branch
* fixed the conflicts with master
* reran the tests, all are still passing

I'm merging this.

@GeorgeJahad Thanks for the help!
